### ISSUE TITLE
Error on releasing mutex

### ISSUE THREAD

**Issue by jehof** [sergeyzwezdin/Hangfire.Mongo#188]
Occasionally we get the error message "Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1"

and and exception like 

> [2019-05-29T12:47:24,646][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>#<JSON::ParserError: unexpected token at '{"fallback":"An Error was logged with message: Error observing notifications
> System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
>    at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
>    at Hangfire.Mongo.DistributedLockMutex.Release(String resource)
>    at Hangfire.Mongo.MongoNotificationObserver.Execute(CancellationToken cancellationToken)","color":"warning","fields":[{"title":"Message","value":"Error observing notifications
> System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
>    at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
>    at Hangfire.Mongo.DistributedLockMutex.Release(String resource)

We are running Hangfire in an testenvironment with a single MongoDB (4.0.5). The MongoDB is running within a docker container.

Is there anything we need to configure, to avoid theses errors? We are using a simple setup like

```
services.AddHangfire(hangfire => hangfire
                .SetDataCompatibilityLevel(CompatibilityLevel.Version_170)
                .UseSimpleAssemblyNameTypeSerializer()
                .UseRecommendedSerializerSettings()
                .UseMongoStorage(configuration.GetConnectionString("Hangfire"), "test", new MongoStorageOptions() { CheckConnection = false, MigrationOptions = new MongoMigrationOptions(MongoMigrationStrategy.Migrate) }));
```

**Comment by gottscj** [comment]
@jehof ,

Would you be able to add some logging?, you could add

GlobalConfiguration.Configuration.UseLogProvider(new ColouredConsoleLogProvider());
I also wrote a FileLogProvider which writes to a file, its here:
https://github.com/sergeyzwezdin/Hangfire.Mongo/blob/master/src/Hangfire.Mongo.Sample.ASPNetCore/FileLogProvider.cs

That would make it easier to check whats happening. The code checking the semaphore is probably a too optimistic.

Thanks!

**Comment by jehof** [comment]
I´ve turned logging on. lets see what happens. I post the results tomorrow.

**Comment by jehof** [comment]
Here a some logs that I have recorded. I hope it helps to track down the problem

{"Timestamp":"2019-06-04T07:36:22.2576129+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:36:48.6491409+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:36:52.2596597+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:37:02.5877241+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1","SourceContext":"Hangfire.Mongo.DistributedLockMutex","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1"}]}}
{"Timestamp":"2019-06-04T07:37:02.6329707+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.MongoNotificationObserver.Execute(CancellationToken cancellationToken)","SourceContext":"Hangfire.Mongo.MongoNotificationObserver","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.Mon...
{"Timestamp":"2019-06-04T07:37:18.6508583+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:37:22.2618231+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:37:48.6531317+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:37:52.2637618+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:38:18.6550410+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:38:22.2656405+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:38:48.6577529+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:38:52.2676691+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:39:18.6595469+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:39:22.2703222+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:39:48.6613169+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:39:52.2719863+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:40:03.4145754+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1","SourceContext":"Hangfire.Mongo.DistributedLockMutex","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1"}]}}
{"Timestamp":"2019-06-04T07:40:03.4250840+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.MongoNotificationObserver.Execute(CancellationToken cancellationToken)","SourceContext":"Hangfire.Mongo.MongoNotificationObserver","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.Mon...
{"Timestamp":"2019-06-04T07:40:18.6630361+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:40:22.2741733+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:40:48.6736508+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:40:52.2762677+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:41:03.8037159+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1","SourceContext":"Hangfire.Mongo.DistributedLockMutex","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error releasing mutex for resource 'Hangfire:recurring-jobs:lock' current count: 1"}]}}
{"Timestamp":"2019-06-04T07:41:03.8048545+00:00","Level":"Error","MessageTemplate":"{State:l}","Properties":{"State":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.MongoNotificationObserver.Execute(CancellationToken cancellationToken)","SourceContext":"Hangfire.Mongo.MongoNotificationObserver","MachineName":"myserver","ThreadId":50},"Renderings":{"State":[{"Format":"l","Rendering":"Error observing notifications\r\nSystem.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.\n   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)\n   at Hangfire.Mongo.DistributedLockMutex.Release(String resource)\n   at Hangfire.Mongo.Mon...
{"Timestamp":"2019-06-04T07:41:18.6755528+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:41:22.2780065+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:85835f4c heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":46},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:85835f4c heartbeat successfully sent"}]}}
{"Timestamp":"2019-06-04T07:41:48.6784821+00:00","Level":"Debug","MessageTemplate":"{State:l}","Properties":{"State":"Server myserver:1:25caed14 heartbeat successfully sent","SourceContext":"Hangfire.Server.ServerHeartbeatProcess","MachineName":"myserver","ThreadId":10},"Renderings":{"State":[{"Format":"l","Rendering":"Server myserver:1:25caed14 heartbeat successfully sent"}]}}

**Comment by gottscj** [comment]
@jehof ,
I apologize, I didnt mention you should set the loglevel = TRACE.

Are you running multiple hangfire instances or are their anything else you can tell me so I can reproduce your environment.

Thanks

**Comment by gottscj** [comment]
@jehof ,

no need, the bug is pretty obvious now that im looking at it.

**Comment by jehof** [comment]
ok. that sounds good. It tried **Trace** as LogLevel but Serilog reported an error.

No that I looked again at the configuration basics of Serilog I should have used **Verbose**. 😄

**Comment by gottscj** [comment]
@jehof ,

The issue should be fixed in 0.6.1. Could you please let me know if you agree. :)

Thanks!

**Comment by jehof** [comment]
The issue seems to be fixed. Good job
